application/src/main/java/org/togetherjava/tjbot/commands/moderation/ReportCommand.java

```diff
@@ -0,0 +1,179 @@
+            logger.warn(
+                    "Cannot find the designated modmail channel in server by id {} with the pattern {}",
+                    guildID, configModMailChannelPattern);
+            return;
+        }
+        InteractionHook hook = event.getHook();
+        event.deferReply().setEphemeral(true).queue();
+        String modalMessage = event.getValue(REPORT_REASON).getAsString();
+        long userID = event.getUser().getIdLong();
+        MessageCreateAction modMessage = createModMessage(modalMessage, userID, reportedMessage,
+                reportedMessageUrl, modMailAuditLog.orElseThrow());
+
+        sendModMessage(hook, modMessage);
+    }
+
+    private boolean handleIsOnCooldown(long userID, MessageContextInteractionEvent event) {
+        if (!isAuthorOnCooldown(userID)) {
+            return false;
+        }
+        event.reply("Can only be used once per %s minutes.".formatted(COOLDOWN_DURATION_VALUE))
+            .setEphemeral(true)
+            .queue();
+        return true;
+    }
+
+    private boolean isAuthorOnCooldown(long userId) {
+        return Optional.ofNullable(authorToLastReportInvocation.getIfPresent(userId))
+            .map(sinceCommandInvoked -> sinceCommandInvoked.plus(COOLDOWN_DURATION_VALUE,
+                    COOLDOWN_DURATION_UNIT))
+            .filter(Instant.now()::isBefore)
+            .isPresent();
+    }
+
+    private Optional<TextChannel> getModMailChannel(JDA jda, long guildID) {
+        return jda.getGuildById(guildID)
+            .getTextChannelCache()
+            .stream()
+            .filter(channel -> modMailChannelNamePredicate.test(channel.getName()))
+            .findAny();
+    }
+
+    private MessageCreateAction createModMessage(String modalMessage, long userID,
+            String reportedMessage, String reportedMessageUrl, TextChannel modMailAuditLog) {
+        MessageEmbed embed = new EmbedBuilder().setTitle("Report")
+            .setDescription(("""
+                    Reported Message: **%s**
+                    [Reported Message URL](%s)
+                    Modal Message: **%s**""").formatted(reportedMessage, reportedMessageUrl,
+                    modalMessage))
+            .setAuthor("Author ID: %s".formatted(userID))
+            .setColor(AMBIENT_COLOR)
+            .setTimestamp(Instant.now())
+            .build();
+
+        return modMailAuditLog.sendMessageEmbeds(embed);
+    }
+
+    private void sendModMessage(InteractionHook hook, MessageCreateAction message) {
+        message.mapToResult().map(result -> {
+            if (result.isSuccess()) {
+                return "Thank you for reporting this message. A moderator will take care of the matter as soon as possible.";
+            }
+            logger.warn("Unable to forward a message report to modmail channel.");
+            return "Sorry, there was an issue sending your report to the moderators. We are investigating.";
+        }).flatMap(hook::editOriginal).queue();
+    }
```

Comment:
The whole mapping thing is a bit unreadable. I would suggest doing `message.queue(successCallback, errorCallback)` and editing the message in both.

Reply:
It was discussed a while ago, on another command that this is the way to do it. Don't know if they still want it this way.

Reply:
![image](https://user-images.githubusercontent.com/54677650/198363767-a228e128-0a14-41d5-8059-8690492e0e69.png)

Don't you think that this is way better?

Reply:
Yes it does, but it does not look like that anymore

Reply:
You just moved things around a bit. Its still weird.

Reply:
Its much better now